docs/developer/decision-records/2022-06-03-event-framework/README.md

```diff
@@ -0,0 +1,94 @@
+
+## Rationale
+
+There are two asynchronous communication levels in EDC:
+
+#### Protocol level
+Every call from a connector to another regards the communication protocol, currently the only protocol implemented in EDC is IDS.
+A protocol call could have an asynchronous response, and that's called **callback**.
+
+#### Instance level
+The EDC should have the possibility to communicate with third party systems (known as subscribers) notifying them about what's happening in the connector. This communication needs to be made through **events**.
+An event is a representation of something that happened in the past, and has very specific characteristics, like:
+- it's named with a short sentence in the past tense (e.g. "contract negotiation confirmed")
+- it's serializable
+- it contains the issue timestamp
+- it describes which layer emitted the event: domain, system, security, and so on...
+
+#### Main differences between Events and Callbacks
+- Callbacks are always correlated with a request, events don't
+- Callbacks are unicast, events are multicast
+- Callbacks are at the protocol level, events can be at a different level
+
+## Approach
+
+An event is generated after a specific state change happens.
+In the EDC will be handled through the `Observable` pattern:
+
+For every entity type, or for other layers a dedicated `Observable` implementation with a dedicated `Listener` should exist.
+The `Listener` interface should have a method for every change that could happen on that entity. For example:
+```java
+public class TransferProcessEventObservableImpl implements Observable<TransferProcessEventListener> {
+    ...
+}
+
+public interface TransferProcessEventListener {
+
+    void created(TransferProcess process);
+
+    void provisioned(TransferProcess process);
+
+    ...
+
+}
+```
+
+The `Listener` would have a single implementation that will take care of generate the event class and pass it to the `EventRouter` component.
+`Event` will be the abstract class that represents all the events, and every event class will extend it.
+
+
+```java
+
+public class TransferProcessEventListenerImpl implements TransferProcessEventListener {
+
+    private final EventRouter eventRouter;
+
+    public void created(TransferProcess process) {
+        var event = TransferProcessCreatedEvent.Builder().newInstance()
+            .id(process.getId())
+            ...
+            .build();
+
+        eventRouter.publish(event);
+    }
+
+    ...
+
+```
+
+The `EventRouter` implementation will be a component that will permit the registration of `EventSubscriber`s and that will dispatch every event to every one of them, asynchronously:
+
+```java
+public interface EventSubscriber {
+    void on(Event event);
+}
+
+public class EventRouterImpl implements EventRouter {
+
+    private List<EventSubscriber> subscribers;
+
+    public void publish(Event event) {
+        subscribers.forEach(subscriber -> {
```

Comment:
Do you think this should be an async future? That will mean events may be lost. We could leave the dispatch choice to the listener implementation, which would provide more flexibility.

Reply:
@jimmarino I think the event framework shouldn't block the normal flow of the EDC, since the event is generated it should be sent to the subscribers without having one blocking the others. 
The event could be lost also if handled in a sync manner

Reply:
With the sync behavior, we could have listeners abort an operation but that may not be desirable. 

I do like the idea of not blocking the normal flow of execution. I think we should document that a side effect of this design is that there is no guarantee events will be delivered (e.g. at least once).

Also. should we use a thread pool for the dispatch? That may be getting too much into implementation detail, but the code sample brought the issue up.

Reply:
I documented this detail in the DR, I think we'd like to reach "at least once" adding persistence in the future, but for the moment this seems a bit out of the scope (and would complicate the feature too much in my opinion).

I think the "thread pool" topic is absolutely an implementation detail, since it's only a matter on how the `runAsync` is called. Currently we have different thread pools spawned in the codebase, I guess we will do pretty much the same (making the pool size configurable)

Reply:
Is there any plan to implement a DLQ in case a event dispatch failed?

Reply:
@bscholtes1A I think an eventual DLQ should be in charge of the dispatcher extension.

Reply:
If I understand it correctly, at the beginning of its existence the Event Framework will not be able to guarantee that all events can really be read from the EventRouter, for example due to an overflow.

Is then the use of this framework for an audit logging usable or are first less critical tasks to be served with it? I don't know which tasks you currently hope to solve with this framework. Of course, this framework offers many application possibilities in general.

Reply:
@janpmeyer every subscriber will receive all the events but, in the case the EDC is stopped after an event emission but before its dispatchment, that event will be lost.
the first use-case we want to accomplish is asynchronous notification. 
For auditing an event persistence will be necessary, but's not part of this proposal, as it is big enough. Will be implemented in the future.